src/Stratis.Bitcoin.Features.BlockStore/BlockRepository.cs

```diff
@@ -139,6 +185,127 @@ public virtual Task InitializeAsync()
             return task;
         }
 
+        /// <inheritdoc />
+        public async Task PruneDatabase(ChainedHeader consensusTip, bool nodeInitializing)
```

Comment:
Instead of calling this on shutdown and initialization I'd also call it from a background thread that would cleanup old blocks every N seconds. Otherwise for pruning to kick in you need to restart the node to reduce how much disk space you are consuming

Reply:
@noescape00 yes @dangershony and I discussed this but wouldn't this mean that we need to introduce a global lock? 

We are worried to make this PR a big one so said that it might be OK to let the node do the shrinking on startup and shutdown only, otherwise we need to makes do the node which might have a muhc deeper impact?

Reply:
I don't think global lock will be needed. Why? 

>We are worried to make this PR a big one so said that it might be OK to let the node do the shrinking on startup and shutdown only, otherwise we need to makes do the node which might have a much deeper impact?

If we do it like that there is little sense to have pruning. I think we need to make pruning work similar to what BTC does.

Reply:
OK so we have to remember that there are 2 concepts here:

**Pruning:**

This deletes the blocks from the store but the file size on disk remains the same.

**Compacting:**

This will recreate the database file on disk without the deleted reference, thus decreasing the file size.

We can't compact the block database during the node's life time as we physically deleting the file on disk and recreating it.. UNLESS we take a lock and wait for it to finish.

Reply:
Yeah. I'd vote against compacting since it's much more time consuming and increases startup or shutdown greatly. 

It's easier for us to change db that actually deletes data then go with smth other than proper pruning.

Reply:
@noescape00 I would it if we can do this, cause then we don't have to worry about startup and shutdown. But a change like this might mean retesting with the testers, which is time we don't have? What do you think.

Reply:
> Yeah. I'd vote against compacting since it's much more time consuming and increases startup or shutdown greatly.

Just FYI, if we keep on pruning during the node's lifetime and compact during startup and shutdown it happens almost instantly.

Reply:
unfortunately compacting is a limitation of debreeze, if we had to do it in a background task it indeed means a global lock (because this will actually copy table around) so an ok solution is shrink onstartup (and fassa also wanted to do it on shutdown)